OutOfSchool/OutOfSchool.BusinessLogic/Services/Elasticsearch/ElasticIndexEnsureCreatedHostedService.cs

```diff
@@ -0,0 +1,64 @@
+﻿using Elastic.Clients.Elasticsearch;
+using Microsoft.Extensions.Options;
+using OutOfSchool.BusinessLogic.Config.Elasticsearch;
+
+namespace OutOfSchool.BusinessLogic.Services.Elasticsearch;
+
+public class ElasticIndexEnsureCreatedHostedService : IHostedService
+{
+    private readonly IServiceProvider services;
+    private readonly ElasticsearchClient client;
+    private readonly string indexName;
+    private readonly int checkConnectivityDelayMs;
+    private readonly int connectionWaitingTimeSec;
+    private readonly ElasticsearchWorkshopConfiguration configurator = new();
+    private readonly ILogger<ElasticIndexEnsureCreatedHostedService> logger;
+
+    public ElasticIndexEnsureCreatedHostedService(
+        IServiceProvider services,
+        ElasticsearchClient client,
+        IOptions<ElasticConfig> elasticOptions,
+        ILogger<ElasticIndexEnsureCreatedHostedService> logger)
+    {
+        this.services = services;
+        this.client = client;
+        this.logger = logger;
+        var config = elasticOptions?.Value
+            ?? throw new ArgumentNullException(nameof(elasticOptions));
+        indexName = config.WorkshopIndexName
+            ?? throw new ArgumentNullException(nameof(elasticOptions), "WorkshopIndexName is null");
+        checkConnectivityDelayMs = config.CheckConnectivityDelayMs;
+        connectionWaitingTimeSec = config.ConnectionWaitingTimeSec;
+    }
+
+    public async Task StartAsync(CancellationToken cancellationToken)
+    {
+        if (cancellationToken.IsCancellationRequested)
+        {
+            logger.LogInformation("Elastic index ensure operation was cancelled");
+            return;
+        }
+
+        var startTime = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
+        using var scope = services.CreateScope();
+        var elasticHealthService = scope.ServiceProvider.GetService<IElasticsearchHealthService>();
+        while (!elasticHealthService.IsHealthy
+                && (DateTimeOffset.UtcNow.ToUnixTimeSeconds() - startTime < connectionWaitingTimeSec))
+        {
+            logger.LogInformation("Waiting for Elastic connection");
+            await Task.Delay(checkConnectivityDelayMs).ConfigureAwait(false);
+        }
+
+        if (elasticHealthService.IsHealthy)
```

Comment:
should not we fail if elastic there is no connection to elastic ?
Logs ?

Reply:
> should not we fail if there is no connection to elastic ?

I discussed this question with @DmyMi, and he suggested adding messages to the log without returning errors, since if Elasticsearch is not working, the application uses a different strategy for search.
I've added logging to indicate when an Elasticsearch index is created successfully and when Elasticsearch index failed to ensure due to a connection timeout.

> If this service is not required - why do we wait for the connection ?

Whether the service is required or not is determined by the options' settings.
If the service is enabled through the options, we wait for a connection in case, for example, Elasticsearch is temporarily unavailable.